partiql-lang/src/main/kotlin/org/partiql/lang/planner/transforms/plan/PlanTyper.kt

```diff
@@ -0,0 +1,1535 @@
+        )
+    }
+
+    //
+    //
+    // EXPRESSIONS
+    //
+    //
+
+    override fun visitRexQueryCollection(node: Rex.Query.Collection, ctx: Context): PlanNode {
+        val input = visitRel(node.rel, ctx)
+        return when (val constructor = node.constructor) {
+            null -> {
+                node.copy(rel = input)
+            }
+            else -> {
+                val typedConstructor = typeRex(constructor, input, ctx)
+                val type = when (input.getProperties().contains(Property.ORDERED)) {
+                    true -> ListType(typedConstructor.grabType() ?: handleMissingType(ctx))
+                    false -> BagType(typedConstructor.grabType() ?: handleMissingType(ctx))
+                }
+                return node.copy(
+                    type = type
+                )
+            }
+        }
+    }
+
+    override fun visitRexPath(node: Rex.Path, ctx: Context): Rex.Path {
+        val ids = grabFirstIds(node)
+        val qualifier = ids.getOrNull(0)?.qualifier ?: Rex.Id.Qualifier.UNQUALIFIED
+        val path = BindingPath(ids.map { rexIdToBindingName(it) })
+        val pathAndType = findBind(path, qualifier, ctx)
+        val remainingFirstIndex = pathAndType.levelsMatched - 1
+        val remaining = when (remainingFirstIndex > node.steps.lastIndex) {
+            true -> emptyList()
+            false -> node.steps.subList(remainingFirstIndex, node.steps.size)
+        }
+        var currentType = pathAndType.descriptor.toStaticType()
+        remaining.forEach { pathComponent ->
+            currentType = when (pathComponent) {
+                is Step.Key -> {
+                    val type = inferPathComponentExprType(currentType, pathComponent, ctx)
+                    type
+                }
+                is Step.Wildcard -> currentType
+                is Step.Unpivot -> error("Not implemented yet")
+            }
+        }
+        return node.copy(
+            type = currentType
+        )
+    }
+
+    override fun visitRexId(node: Rex.Id, ctx: Context): Rex.Id {
+        val bindingPath = BindingPath(listOf(rexIdToBindingName(node)))
+        return node.copy(type = findBind(bindingPath, node.qualifier, ctx).descriptor.toStaticType())
+    }
+
+    override fun visitRexBinary(node: Rex.Binary, ctx: Context): Rex.Binary {
+        val lhs = visitRex(node.lhs, ctx).grabType() ?: handleMissingType(ctx)
+        val rhs = visitRex(node.rhs, ctx).grabType() ?: handleMissingType(ctx)
+        val args = listOf(lhs, rhs)
+        val type = when (node.op) {
+            Rex.Binary.Op.PLUS, Rex.Binary.Op.MINUS, Rex.Binary.Op.TIMES, Rex.Binary.Op.DIV, Rex.Binary.Op.MODULO -> when (hasValidOperandTypes(args, node.op.name, ctx) { it.isNumeric() }) {
+                true -> computeReturnTypeForNAry(args, PlanTyper::inferBinaryArithmeticOp)
+                false -> StaticType.NUMERIC // continuation type to prevent incompatible types and unknown errors from propagating
```

Comment:
I likely need to read the code in more depth but can you please elaborate on this?

Reply:
Essentially the same as [here](https://github.com/partiql/partiql-lang-kotlin/blob/965183a2b65d944acc541ad7e0f88929332680f6/partiql-lang/src/main/kotlin/org/partiql/lang/eval/visitors/StaticTypeInferenceVisitorTransform.kt#L447-L495).

For binary arithmetic operations, we make sure both operands are numeric, then we compute the return type using the inferBinaryArithmeticOp.